File: 03/03.py
def coords_for_path(path):
    x, y = 0, 0
    coords = []
    for corner in path:
        direction = corner[0]
        length = int(corner[1:])

        dx, dy = 0, 0
        if direction == 'R':
            dx = 1
        if direction == 'L':
            dx = -1
        if direction == 'U':
            dy = 1
        if direction == 'D':
            dy = -1

        for _ in range(length):
            x += dx
            y += dy
            coords.append((x, y))
    return coords


def parse_file(file):
    with open(file) as f:
        path1 = f.readline().split(',')
        path2 = f.readline().split(',')
    return path1, path2


def distance(x, y):
    return abs(x) + abs(y)


def lowest_distance(coords):
    lowest = float('inf')
    for x, y in coords:
        d = distance(x, y)
        if d < lowest:
            lowest = d
    return lowest


def lowest_step(intersection, coords1, coords2):
    lowest_step = float('inf')
    for x, y in intersections:
        first = coords1.index((x, y))
        second = coords2.index((x, y))
        steps = first + second + 2
        lowest_step = min(steps, lowest_step)
    return lowest_step


path1, path2 = parse_file('input.path')

coords1 = coords_for_path(path1)
coords2 = coords_for_path(path2)

intersections = set(coords1) & set(coords2)

print('Part 1:', lowest_distance(intersections))
print('Part 2:', lowest_step(intersections, coords1, coords2))

File: 03/03.clj
(ns dec03)
  (require '[clojure.string :as string])
  (require '[clojure.set :as sets])

(defn parse-int [n] (Integer/parseInt (str n)))

(defn parse-file [file]
  (->> (slurp file)
       (string/split-lines)
       (map #(string/split % #","))))

(defn direction->deltas [d]
  (case d
    \R [1 0]
    \L [-1 0]
    \U [0 1]
    \D [0 -1]))

(defn parse-direction [d]
  [(direction->deltas (first d))
   (parse-int (subs d 1))])

(defn add-coordinates [a b]
  [(+ (first a) (first b) )
   (+ (second a) (second b))])

(defn distance [[x y]] (+ (Math/abs x) (Math/abs y)))

(defn step->coordinates [coordinates [delta steps]]
  (loop [n 0
         coords coordinates
         current (add-coordinates (last coordinates) delta)]
    (if (= n steps)
      coords
      (recur (inc n)
             (conj coords current)
             (add-coordinates current delta)))))

(defn path->coordinates [path]
  (reduce step->coordinates [[0 0]] path))

(defn row->coords [r]
  (->> (map parse-direction r)
       (path->coordinates)
       (set)))

(println "Part 1:" 
  (->> (parse-file "input.path")
       (map row->coords)
       (reduce sets/intersection)
       (map distance)
       (set)
       (#(disj % 0))
       (reduce min)))